// revolt.js/dist/api/permissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_PERMISSION_DM = exports.U32_MAX = exports.ServerPermission = exports.ChannelPermission = exports.UserPermission = void 0;
exports.UserPermission = {
    Access: 0b00000000000000000000000000000001,
    ViewProfile: 0b00000000000000000000000000000010,
    SendMessage: 0b00000000000000000000000000000100,
    Invite: 0b00000000000000000000000000001000, // 8
};
exports.ChannelPermission = {
    View: 0b00000000000000000000000000000001,
    SendMessage: 0b00000000000000000000000000000010,
    ManageMessages: 0b00000000000000000000000000000100,
    ManageChannel: 0b00000000000000000000000000001000,
    VoiceCall: 0b00000000000000000000000000010000,
    InviteOthers: 0b00000000000000000000000000100000,
    EmbedLinks: 0b00000000000000000000000001000000,
    UploadFiles: 0b00000000000000000000000010000000, // 128
};
exports.ServerPermission = {
    View: 0b00000000000000000000000000000001,
    ManageRoles: 0b00000000000000000000000000000010,
    ManageChannels: 0b00000000000000000000000000000100,
    ManageServer: 0b00000000000000000000000000001000,
    KickMembers: 0b00000000000000000000000000010000,
    BanMembers: 0b00000000000000000000000000100000,
    ChangeNickname: 0b00000000000000000001000000000000,
    ManageNicknames: 0b00000000000000000010000000000000,
    ChangeAvatar: 0b00000000000000000100000000000000,
    RemoveAvatars: 0b00000000000000001000000000000000, // 32768
};
exports.U32_MAX = Math.pow(2, 32) - 1; // 4294967295
exports.DEFAULT_PERMISSION_DM = exports.ChannelPermission.View
    + exports.ChannelPermission.SendMessage
    + exports.ChannelPermission.ManageChannel
    + exports.ChannelPermission.VoiceCall
    + exports.ChannelPermission.InviteOthers
    + exports.ChannelPermission.EmbedLinks
    + exports.ChannelPermission.UploadFiles;
